go-controller/pkg/node/egressfirewall_dns/egressfirewall_dns.go

```diff
@@ -0,0 +1,306 @@
+func (e *EgressDNS) Update(dns string) (bool, error) {
+	return e.dns.Update(dns)
+}
+
+// Run() spawns go-routines that handles add/update/delete of dns entries for dnsNames used in
+// EgressFirewalls.
+func (e *EgressDNS) Run(defaultInterval time.Duration) {
+	go e.addDNSNames()
+	go e.updateDNSRecord(defaultInterval)
+
+}
+
+// go-routine that handles adding new DNS names to the resolver, it is signaled by Add()
+// which is in turn is triggered by the addition of an egressFirewall with DNS names.
+// Signals the updateDNSRecord go-routine when it is finished adding a DNS so the
+// durationTillNextQuery can be updated if needed
+func (e *EgressDNS) addDNSNames() {
+	for {
+		select {
+		case dnsName := <-e.added:
+			func() {
+				e.lock.Lock()
+				defer e.lock.Unlock()
+				if err := e.dns.Add(dnsName); err != nil {
+					utilruntime.HandleError(err)
+				}
+				if err := e.updateEntryForName(dnsName); err != nil {
+					utilruntime.HandleError(err)
+				}
+				e.signalAddFinished()
+			}()
+		case <-e.stopChan:
+			return
+		case <-e.controllerStop:
+			return
+		}
+	}
+}
+
+// go-routine that handles periodic requiering of the dnsResolver to update
+// the IP addresses that dnsNames resolve to.
+// at least 1 DNS record gets updated every 30 min in the case that there is no
+// TTL less then 30 minutes. Everytime an update occurs only one DNS record is
+// updated. This go-routine is signaled to recompute the next query time after an
+// add operation is completed.
+func (e *EgressDNS) updateDNSRecord(defaultInterval time.Duration) {
+	// initially set the next DNS Query to happen at the default interval
+	durationTillNextQuery := defaultInterval
+	var ttl time.Time
+	var timeSet bool
+	var dnsName string
+	for {
+		select {
+		case <-time.After(durationTillNextQuery):
+			func() {
+				e.lock.Lock()
+				defer e.lock.Unlock()
+				if len(dnsName) > 0 && e.dnsEntries[dnsName] != nil {
+					if _, err := e.Update(dnsName); err != nil {
+						utilruntime.HandleError(err)
+					}
+					if err := e.updateEntryForName(dnsName); err != nil {
+						utilruntime.HandleError(err)
+					}
+				}
+			}()
+		case <-e.addFinished:
+			// after a dnsName gets added this go-routine needs to recompute how long
+			// to wait before waking again
+		case <-e.stopChan:
+			return
+		case <-e.controllerStop:
+			return
+		}
+
+		// before waiting on the signals get the next time this go-routine  needs to wake up
+		ttl, dnsName, timeSet = e.dns.GetNextQueryTime()
+		if time.Until(ttl) > defaultInterval || !timeSet {
+			durationTillNextQuery = defaultInterval
+		} else {
+			durationTillNextQuery = time.Until(ttl)
+		}
+
+	}
+}
+
+func (e *EgressDNS) Remove(dnsNames []string, namespace string) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	resultErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// preemptivly get the NodeDNSInfo so that all the Removes can be processed at once
+		nodeDNSInfo, err := e.wf.GetNodeDNSInfo(e.nodeName)
```

Comment:
I believe you need to DeepCopy this here.

Reply:
Why? when is deepcopy needed and how is it different?